### PR TITLE
Refactored editor

### PR DESCRIPTION
**Pull request by mitchellrj**
Refactored rich-text editor support to make it possible for ``hallo.js`` to be replaced. For example, by [isotoma/wagtailtinymce](https://github.com/isotoma/wagtailtinymce).

I'm confident that I've got this refactored in such a way that doesn't affect the integrity of the current codebase or the interdependencies between packages. The only downside is that users will need to add an extra app to their ``INSTALLED_APPS`` setting.

### CODE REVIEW

**Comment by kaedroho** [comment]
Could we call the module ``wagtailhalloeditor``?

**Comment by tomdyson** [comment]
@JoshBarr do you have any time to review this?

**Comment by mitchellrj** [comment]
@kaedroho renamed.

**Comment by JoshBarr** [comment]
@tomdyson just saw your email about reviewing this. Will take a look.

**Comment by SalahAdDin** [comment]
Had you put a hallo in a separate package? How work it?

**Comment by mitchellrj** [comment]
@SalahAdDin the Hallo editor is now in a separate Django app within the
Wagtail package. Instead of being loaded by default by the editor JS, it
uses the editor JS hooks to load itself if it is in the INSTALLED_APPS.
This means that it can be switched out more easily of a different rich text
editor is preferred. Does that answer your questions?

**Comment by mitchellrj** [comment]
The latest changes allow the editor to be configured in settings, or overridden explicitly on the Field class. This defaults to Hallo.

There is unfortunately no way around the fact that `wagtail.wagtailhalloeditor` needs to be added to `INSTALLED_APPS`. The only way to avoid this would be to merge it back into an app that is already in there. I think there is more merit to keeping it as a separate app within Wagtail and so have not made this change.

**Comment by mitchellrj** [comment]
What do you reckon, @tomdyson, @gasman, @kaedroho?

**Comment by tomdyson** [comment]
We like it, @mitchellrj! We just want to see if we can solve the backwards compatibility issue of requiring `wagtail.wagtailhalloeditor` in `INSTALLED_APPS`. @kaedroho is looking into it.

**Comment by mitchellrj** [comment]
Whichever way I slice it, I can't find a way to programmatically add it to `INSTALLED_APPS` at runtime. There are [hacky] ways to have it added and working during `runserver`, but not in a way that works for static content (notably during `collectstatic`).

File: wagtail/wagtailhalloeditor/editor.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import json
+
+from wagtail.wagtailcore.fields import BaseRichTextEditor
+
+
+class HalloEditor(BaseRichTextEditor):
```
Comment (kaedroho) [inline wagtail/wagtailhalloeditor/editor.py]: Could we move this to ``wagtailadmin/halloeditor.py`` (and move the static files to somewhere like ``wagtail/wagtailadmin/static_src/wagtailadmin/halloeditor/``)?

Perhaps we should also move the ``insert_editor_js/css`` hooks into methods of the editor class as well to make it easier to kick out hallo js.

**Comment by kaedroho** [comment]
Perhaps it might be nice to treat editors in a similar way to how we treat search/cache/database backends and allow them to be configured in settings.

For example:
```
WAGTAILADMIN_RICHTEXT_EDITORS = {
    'default': {
        'EDITOR': 'wagtail.wagtailadmin.halloeditor.HalloEditor',
        .. extra configuration here, eg icons.
    },
    'tinymce': {
        'EDITOR': 'wagtailtinymce.editor.TinyMCEEditor',
    }
}
```

Then on the rich text field itself, you can choose which one to use (this would obviously default to 'default'):
```
RichTextField(editor='tinymce')
```

File: wagtail/wagtailcore/fields.py [hunk 2]
```diff
@@ -1,18 +1,69 @@
...
+from django.db import models
+from django.utils.module_loading import import_string
+from django.utils.six import with_metaclass, string_types, reraise
 
 from wagtail.wagtailcore.rich_text import DbWhitelister, expand_db_html
 from wagtail.utils.widgets import WidgetWithScript
 from wagtail.wagtailcore.blocks import Block, StreamBlock, StreamValue, BlockField
 
 
+class BaseRichTextEditor(object):
```
Comment (kaedroho) [inline wagtail/wagtailcore/fields.py]: Not sure if this should be in wagtailcore or wagtailadmin.. ping @gasman

**Comment by kaedroho** [comment]
Thanks @mitchellrj!

Our main concern here is the need to add that extra app. I think it would be better to model this on how we manage search backends (and previously image processing backends). By moving all the logic around an editor into a single, importable class.

By the looks of it, the reason why ``halloeditor`` is it's own app is to prevent its ``insert_editor_js/css`` hooks being called. I think these hooks would be best moved into methods on the editors class itself as this would allow Wagtail to decide which editors css/js to use based on the ``RichTextField``s on the model being edited (so it can omit the hallo editors js/css if the page being edited only needs tinymce, for example). Using hooks makes it impossible for Wagtail to tell which editor the assets belong to.

**Comment by SalahAdDin** [comment]
:+1:

File: wagtail/wagtailcore/fields.py (hunk 2 above)
Comment (kaedroho) [inline wagtail/wagtailcore/fields.py]: Just discussed with Matthew. He said the following:
 - If the code is for the rich editor interface, it should be in ``wagtailadmin``
 - If the code is for displaying rich text on the frontend, it should be in ``wagtailcore``

So this probably should be moved to ``wagtailadmin``

**Comment by tomdyson** [comment]
Has anyone had a look at https://github.com/basecamp/trix?

> Trix sidesteps these inconsistencies by treating `contenteditable` as an I/O device: when input makes its way to the editor, Trix converts that input into an editing operation on its internal document model, then re-renders that document back into the editor.

**Comment by JoshBarr** [comment]
Trix looks interesting. I agree that working with `contenteditable` itself is painful. 

I'm doing an interesting project with [Quill](https://github.com/quilljs/quill) which has a very nice API, is written in ES6, and internally uses the fantastic rich-text library's [Delta](https://github.com/ottypes/rich-text) format for dealing with complex text operations. The nice thing about Quill is that you can save the Delta (a JSON object) back to your server rather than the HTML (icky). Good when you start thinking about COPE systems and the need to mark up content agnostic of the presentation format.

**Comment by shredding** [comment]
We need this feature urgently.

Is there anything we can do to assist you / test / write additional code to get merged soon?

**Comment by gasman** [comment]
Hi @shredding - the blocker right now is the need for adding wagtailhalloeditor as an extra app. My feeling is that the wagtail_hooks mechanism is probably the wrong tool for offloading bits of *required* core functionality to external modules - I think @kaedroho's suggestion of a switchable backend setting <https://github.com/torchbox/wagtail/pull/1521#issuecomment-137001133> like search / cache / database is a better fit. I'll be able to give this some more thought once 1.2 is out of the way (end of next week, hopefully).

**Comment by mitchellrj** [comment]
I will come back to this shortly, but some other work is taking priority right now. Sorry for the delay in addressing the outstanding issues.

**Comment by tomdyson** [comment]
@shredding are you able to help with the blocker that @gasman explains above?

**Comment by shredding** [comment]
I have not looked in the way this is implemented, but it's not a blocker in terms of "fix this and it will work", it's a blocker that requires an architectonical descision.

We have now used a very downstripped approach at our company, via the fields and I personally think that it may be a better approach.

We've just implemented a "TinyMCEBlock" which you can use instead of the RichTextBlock, and it renders TinyMCE. I don't think that a frontend-editing JS Library should result in a dedicated app in the application backend. You can give people more freedom if they can just switch the editor for what fits more for the current type (e.g. hallo for the basics, tiny for more complex models).

My proposed solution would therefore an argument like `editor` and call it day. We could share our solution if this is of interest.

```python
class ContentPage(Page):
    main_part = StreamField([
        ('heading', blocks.CharBlock()),
        ('simple_paragraph', RichTextBlock()),
        ('complex_paragraph', RichTextBlock(editor='tinymce')),
    ], null=True, blank=True)
```

**Comment by tomdyson** [comment]
Thanks, @shredding. Please do share your solution.

**Comment by janpleines** [comment]
You can find the solution @shredding mentioned here: https://github.com/janpleines/wagtail_tinymce
It's a quick integration of tinymce for our needs and works so far without changing any core code (beside the _editor_js.html).

**Comment by JoshBarr** [comment]
Bump on this one. 

What do we want to do? I've been experimenting with Quill and Trix lately. It would be nice to have a `JSONRichTextField` with a pluggable editor, so we can store serialised data instead of HTML. It would remove the need for quite a few methods that deal with securing and whitelisting HTML elements. 

Trix implements a `element.editor.toJSON()` method, and Quill has `editor.getContents()` to return a chunk of data we can store. Unfortunately they don't use the same representation of the document model under the hood. Of the two, I think Trix might actually be better suited to Wagtail, since it has good RTL support and is well tested in production by Basecamp. 

Only downside I can think of is that it's in coffeescript. That can be mitigated via Webpack loaders, though.

**Comment by m1kola** [comment]
From point of generalization of editor, you may be interested in #2175 ;)

**Comment by jaydensmith** [comment]
+1

**Comment by gasman** [comment]
An alternative approach to this has now been implemented in #2416.

**Pull request closed without merging**